### PR TITLE
CompasHPC improvements

### LINKED ISSUE #12: Move enhancement STROOPWAFEL from Legacy COMPAS to new COMPAS
this is for me

### LINKED ISSUE #18: double check STROOPWAFEL works in newCOMPAS

### LINKED ISSUE #154: Test compatibility of CompasHPC and BSE_Grid.txt
Users can specify a list of binaries to evolve using a `BSE_grid.txt` file. 

I suspect that this may not work as intended in conjunction with CompasHPC.

My concern is that if the user runs a grid file of N_lines run on HPC using N_batches, presumably the intended result is for each batch to run N_lines/N_batches (in order to speed up the run). Currently, I think you would get N_batches versions of the same run. I haven't tested this though, but unless someone else has implemented the splitting of the file, I think this will be an issue.

### PR DESCRIPTION
**Pull request by reinhold-willcox**
Changes related to CompasHPC and stroopwafel sampling. Also, to pythonSubmitDefault.py.

Removed CompasHPC directory - all functionality now in `defaults/stroopwafel_interface_default.py`.

`defaults/stroopwafel_interface_default.py` runs an interface to stroopwafel, so will sample over a list of input parameters to produce grid files, then run these grids across a specified number of cores. Need to run `pip install stroopwafel` first. 

The file works on its own, but if there is a `pythonSubmit.py` in the same dir, and the appropriate flag has been set, `interface.py` will run with the options in the pythonSubmit (i.e combining stroopwafel sampling with pythonSubmit list of options).

In order to combine them, the structure of pythonSubmit has been changed so that compas arguments are now collected into a dictionary. This doesn't change how most users interact with it, but it means the "hypergrid" and "hyperlist" functionality are no longer supported. I know some people wanted to keep this for the time being, until stroopwafel sampling is more mainstream, so the old file has been renamed to "pythonSubmitDefault_old.py". 

Works also on ozstar, but you'll need to run `pip install stroopwafel --user` there first. 

Note, the "Adaptive Importance" part of stroopwafel is still not working, that will come later.

Documentation added in `docs/sampling.md` and `docs/compasHPC.md`.

### CODE REVIEW

**Review by jeffriley — COMMENTED** [review]
The changes look good to me, but I know next to nothing about OzStar or the Stroopwafel implementation, especially Loki's recent work - I'll leave the decision to approve to Loki and Simon.

**Comment by reinhold-willcox** [comment]
Ok, cool. It currently conflicts because of the changelog, but I'll wait for approval from the maestro's  before updating.

**Comment by ilyamandel** [comment]
Pinging @lokiysh , @SimonStevenson , @FloorBroekgaarden .

File: defaults/stroopwafel_interface_default.py
Comment (lokiysh) [inline defaults/stroopwafel_interface_default.py]: Do we really require random seed here? Wont it overwrite the one in pythonSubmit ?

File: defaults/stroopwafel_interface_default.py
Comment (lokiysh) [inline defaults/stroopwafel_interface_default.py]: There is no need to run the adaptation phase if we are not running the refinement phase. So we can comment out this line as well, it will unnecessarily use up resources

File: docs/sampling.md
Comment (lokiysh) [inline docs/sampling.md]: Good to add here is that debug is specifically used to print the output / error that COMPAS will sendover, which is otherwise hidden. It has nothing to do with debugging stroopwafel.

File: defaults/pythonSubmitDefault.py
Comment (lokiysh) [inline defaults/pythonSubmitDefault.py:604]: I assume with this we are still keeping the ability to run pythonSubmit independently ?

File: defaults/pythonSubmitDefault.py
Comment (lokiysh) [inline defaults/pythonSubmitDefault.py:517]: nothing particularly bad, but just wondering what changed the indentation? If it works, it works!

File: defaults/stroopwafel_interface_default.py
Comment (lokiysh) [inline defaults/stroopwafel_interface_default.py]: This should not be hardcoded here! What if the user tries to change it from pythonSubmit, its not being respected.

File: defaults/stroopwafel_interface_default.py
Comment (lokiysh) [inline defaults/stroopwafel_interface_default.py]: better to put this at the top of the file, since we are saying somewhere in the file "Do not change below or something".

File: defaults/stroopwafel_interface_default.py
Comment (reinhold-willcox) [reply defaults/stroopwafel_interface_default.py]: It doesn't quite work to leave the random seed in from the pythonSubmit, because the same seed is passed to each new batch (i.e for each call of COMPAS). It's not a huge problem, since the grids should all be different (which covers most of what was previously handled by the random number), but any numbers that are pulled purely from a random draw will be repeated. 

A possible fix to this would be use the random-seed from pythonSubmit, but incremented by the number of binaries in a given batch times the index of the batch. I'll give that a go and see if I run into any problems.

File: defaults/pythonSubmitDefault.py
Comment (reinhold-willcox) [reply defaults/pythonSubmitDefault.py:517]: By redefining the function as a method of the pythonProgramOptions class, it made it easier to import only the class into the stroopwafel interface.

File: defaults/pythonSubmitDefault.py
Comment (reinhold-willcox) [reply defaults/pythonSubmitDefault.py:604]: Yes, the internals are different, but to most users the pythonSubmit can still be configured and run like previously. The only difference is the removal of the hyper-parameter grids, which hopefully will be implementable using the new sw interface.

File: defaults/stroopwafel_interface_default.py
Comment (reinhold-willcox) [reply defaults/stroopwafel_interface_default.py]: Ya I agree there's a bit of a degeneracy here.

Ultimately, it's always going to be a little ambiguous which file wins out in the scenario where sw and pySub are both used. But in this case, since people would be actually running the sw_interface file, I figured it makes more sense to accept the sw value in stead of the pySub value. 

I'm open to discussion / disagreement on this, but in the meantime I'll try to make it clearer in the comments.

File: defaults/stroopwafel_interface_default.py
Comment (reinhold-willcox) [reply defaults/stroopwafel_interface_default.py]: Fair point

File: docs/sampling.md
Comment (reinhold-willcox) [reply docs/sampling.md]: Oh really? I didn't know that, thanks!

File: defaults/stroopwafel_interface_default.py
Comment (reinhold-willcox) [reply defaults/stroopwafel_interface_default.py]: agreed, that's been fixed

File: defaults/stroopwafel_interface_default.py
Comment (reinhold-willcox) [reply defaults/stroopwafel_interface_default.py]: this has now been fixed

**Review by lokiysh — APPROVED** [review]

**Pull request merged**
